addon-test-support/@ember/test-helpers/dom/fire-event.ts

```diff
@@ -88,48 +95,57 @@ function fireEvent(
   element: Element | Document | Window,
   eventType: string,
   options = {}
-): Event | void {
-  if (!element) {
-    throw new Error('Must pass an element to `fireEvent`');
-  }
-
-  let event;
-  if (isKeyboardEventType(eventType)) {
-    event = _buildKeyboardEvent(eventType, options);
-  } else if (isMouseEventType(eventType)) {
-    let rect;
-    if (element instanceof Window && element.document.documentElement) {
-      rect = element.document.documentElement.getBoundingClientRect();
-    } else if (isDocument(element)) {
-      rect = element.documentElement!.getBoundingClientRect();
-    } else if (isElement(element)) {
-      rect = element.getBoundingClientRect();
-    } else {
-      return;
-    }
-
-    let x = rect.left + 1;
-    let y = rect.top + 1;
-    let simulatedCoordinates = {
-      screenX: x + 5, // Those numbers don't really mean anything.
-      screenY: y + 95, // They're just to make the screenX/Y be different of clientX/Y..
-      clientX: x,
-      clientY: y,
-      ...options,
-    };
-
-    event = buildMouseEvent(eventType, simulatedCoordinates);
-  } else if (
-    isFileSelectionEventType(eventType) &&
-    isFileSelectionInput(element)
-  ) {
-    event = buildFileEvent(eventType, element, options);
-  } else {
-    event = buildBasicEvent(eventType, options);
-  }
-
-  element.dispatchEvent(event);
-  return event;
+): Promise<Event | void> {
+  return Promise.resolve()
+    .then(() => runHooks('fireEvent', 'start', element))
+    .then(() => runHooks(`fireEvent:${eventType}`, 'start', element))
```

Comment:
I like the flexibility this provides. It's a nice composition of general to specific!